### DISCUSSION TITLE
Adding new functionality to the library

### DISCUSSION THREAD

**Discussion by Ghyath-Serhal** [quartznet/quartznet#2538] [Ideas]
I am currently deploying a Quartz application on AKS, where a single node is handling 4 concurrent Quartz jobs by default.
Occasionally, there are a number of misfired jobs that remain in a queue, waiting for an available thread to execute.

To optimize the application's performance and scale it effectively, I need to track the number of misfired jobs and their corresponding wait times.
For instance, if there are five misfired jobs with wait times exceeding 15 minutes, I would like to trigger the creation of an additional node to handle the load.

From my investigation, it seems that Quartz does not currently provide the ability to track the duration a job has been in a misfired state. I am planning to add this functionality to the Quartz library.

@lahma what are your thoughts?

**Comment by lahma** [comment]
v4 branch has had some OpenTelemetry improvements. I wonder if they would be they way to go. Like reporting usage per scheduler and if you can aggregate information from all schedulers and see that average load is near 100% (thread pool usage) or something similar that would trigger scaling out.

**Reply by Ghyath-Serhal** [reply]
I was thinking of adding a new column to the qrtz_triggers table that contains the initial fire time.

**Reply by lahma** [reply]
What is initial fire time in this context? Next fire time should tell how much "we are late from firing" at the moment. Of course need to check whether trigger is already in processing (and could take multiple minutes to complete). Like in Azure I think reliable metrics would signal about resource exhaustion.

**Reply by Ghyath-Serhal** [reply]
The problem with the next fire time is that it is updated every time there is a misfire, right?

What i mean is if the job gets misfired 5 times before being fired, the next fire time is updated 5 times, and i am not able to know the initial fire time. Hence i am not able to know how long the job is late.

The initial fire time will not be updated on misfire.

**Reply by lahma** [reply]
Fair enough, but wouldn't constant 100% thread pool utilization tell you the same story that there's too much load?

**Reply by Ghyath-Serhal** [reply]
Indeed, thread pool utilization could offer some insight, yet this data often lacks precision. Consider a situation where there is a high load, but no jobs are misfiring; in such cases, scaling out would be ineffective.

Additionally, it's possible that the load is on the database server rather than on the application server.

Therefore, the most effective approach for me would be to monitor the number of misfiring jobs and the duration of their delays in order to be able to take correct decisions about scaling

**Comment by Ghyath-Serhal** [comment]
Hello @lahma, I have created a separate branch and implemented the change on it, can you please add me as a collaborator to the repo in order to push the changes.
Thank you

**Reply by lahma** [reply]
I don't give contributor access for code proposals. See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/best-practices-for-pull-requests , there's guidance what good PRs are and how they are done. I'm still not convinced the database schema path is something I want to accept.

**Reply by Ghyath-Serhal** [reply]
Hello.

Could you please take a look when you have some time? I’d really appreciate your feedback and any comments or suggestions you might have.

**https://github.com/quartznet/quartznet/pull/2543**
